packages/app/src/app/overmind/internalActions.ts

```diff
@@ -590,61 +591,114 @@ export const setViewModeForDashboard = ({ effects, state }: Context) => {
   }
 };
 
-export const setActiveTeamFromUrlOrStore = async ({ actions }: Context) =>
-  actions.internal.setActiveTeamFromUrl() ||
-  actions.internal.setActiveTeamFromLocalStorage() ||
-  actions.internal.setActiveTeamFromPersonalWorkspaceId();
+const INVALID_ID_TITLE = 'Workspace not recognized.';
+const INVALID_ID_MESSAGE =
+  "The workspace in the URL or stored in your browser is unknown. We've automatically switched to your personal workspace.";
 
-export const setActiveTeamFromLocalStorage = ({
+// Rename to initializeTeam (because we also use personalWorkspaceId);
+export const setActiveTeamFromUrlOrStore = async ({
+  actions,
+  effects,
+}: Context) => {
+  const { id, isValid } = await actions.internal.getTeamIdFromUrlOrStore();
+
+  if (isValid) {
+    // Set active team from url or storage.
+    actions.setActiveTeam({ id });
+  } else {
+    // If an id was set but it's not valid we show a toast message informing the user
+    // we changed the workspace to the personal workspace. If no id was set we silently
+    // activate the personal team.
+    if (id) {
+      effects.notificationToast.add({
+        title: INVALID_ID_TITLE,
+        message: INVALID_ID_MESSAGE,
+        status: NotificationStatus.NOTICE,
+      });
+    }
+
+    // Change to personal workspace.
+    actions.internal.setActiveTeamFromPersonalWorkspaceId();
+  }
+};
+
+export const getTeamIdFromUrlOrStore = async ({
+  state,
   effects,
   actions,
 }: Context) => {
-  const localStorageTeam = effects.browser.storage.get(TEAM_ID_LOCAL_STORAGE);
+  const suggestedTeamId =
+    actions.internal.getTeamIdFromUrl() ||
+    actions.internal.getTeamIdFromLocalStorage();
+  const hasTeams = state.dashboard?.teams && state.dashboard.teams.length > 0;
 
-  if (typeof localStorageTeam === 'string') {
-    actions.setActiveTeam({ id: localStorageTeam });
-    return localStorageTeam;
+  let userTeams: TeamFragmentDashboardFragment[] | undefined;
+
+  if (hasTeams) {
+    userTeams = state.dashboard.teams;
+  } else {
+    // TODO: Instead of actions.dashboard.getTeams();
+    // We might be able to use it though, and then use state.dashboard.teams!
+    const teams = await effects.gql.queries.getTeams({});
+
+    if (teams?.me) {
+      userTeams = teams.me.workspaces;
+
+      // Also set state while we're at it
+      state.dashboard.teams = teams.me.workspaces;
+      state.personalWorkspaceId = teams.me.personalWorkspaceId;
+    }
   }
 
-  return null;
+  const isSuggestedTeamValid = userTeams?.some(
+    team => team.id === suggestedTeamId
+  );
+
+  return {
+    id: suggestedTeamId,
+    isValid: isSuggestedTeamValid,
+  };
```

Comment:
What do you think about changing the return?

```suggestion
  return isSuggestedTeamValid ? suggestedTeamId : null;
```

Because the id isn't useful unless it's valid.

Reply:
Nice that this stands out. I have been thinking about this change for a bit when I made it. I thought it made sense to always return the `id` no matter what because the function was called `getTeamIdFromUrlOrStore`. But I think you're right, we don't use it anywhere so I think it's safe to apply your suggestion.